src/object/cli_obj.c

```diff
@@ -2272,16 +2273,24 @@ obj_comp_cb(tse_task_t *task, void *data)
 		 * shard, such as -DER_TIMEDOUT or daos_crt_network_error().
 		 */
 
-		if (!obj_auxi->spec_shard ||
+		if ((!obj_auxi->spec_shard && !obj_auxi->no_retry) ||
 		    task->dt_result != -DER_INPROGRESS)
 			obj_auxi->io_retry = 1;
 
 		if (task->dt_result == -DER_CSUM) {
-			if (!obj_auxi->spec_shard &&
-			    obj_auxi->opc == DAOS_OBJ_RPC_FETCH)
-				obj_auxi->csum_retry = 1;
-			else
+			if (!obj_auxi->spec_shard && !obj_auxi->no_retry &&
+			    obj_auxi->opc == DAOS_OBJ_RPC_FETCH) {
+				if (!obj_auxi->csum_retry &&
+				    !obj_auxi->csum_report) {
+					obj_auxi->csum_report = 1;
+				} else if (obj_auxi->csum_report) {
+					obj_auxi->csum_report = 0;
+					obj_auxi->csum_retry = 1;
+				}
```

Comment:
maybe this is the reason why you remove the reset of csum_retry flag (to only set the csum_report one time?), maybe can be resolved by other method.

the L2283 ~ L2289 seems can be simplified as:
obj_auxi->csum_retry = 1;
obj_auxi->csum_report++;
(change csum_report from a flag to an int, and only set the ORF_CSUM_REPORT when csum_report == 1)